reasoner/planner/RecursivePlanner.java

```diff
@@ -0,0 +1,383 @@
+
+        for (OrderingChoice bestPlanForCall : bestPlan.orderingChoices.values()) {
+            planCache.put(bestPlanForCall.callMode, new Plan(bestPlanForCall.ordering, Math.round(Math.ceil(bestPlan.cost(bestPlanForCall.callMode)))));
+            cyclicScalingFactors.put(bestPlanForCall.callMode, bestPlan.cyclicScalingFactorSum.get(bestPlanForCall.callMode));
+        }
+    }
+
+    private ComponentPlan componentPlanSearch(CallMode root, Set<CallMode> pendingCallModes, Map<CallMode, OrderingChoice> choices) { // The value contains the scaling factors needed for the other conjunctions in the globalCost.
+        // Pick a choice of ordering, expand dependencies, recurse ; backtrack over choices
+        if (pendingCallModes.isEmpty()) { // All modes have an ordering chosen -> we have a complete candidate plan
+            return createComponentPlan(choices);
+        }
+
+        CallMode mode = iterate(pendingCallModes).next();
+        pendingCallModes.remove(mode);
+        assert !choices.containsKey(mode); // Should not have been added to pending
+
+        ConjunctionNode conjunctionNode = conjunctionGraph.conjunctionNode(mode.conjunction);
+        ComponentPlan bestPlan = null; // for the branch of choices committed to so far.
+        double bestPlanCost = Double.MAX_VALUE;
+        for (OrderingChoice orderingChoice : candidateOrderings.get(mode)) {
+            choices.put(mode, orderingChoice);
+
+            Set<CallMode> nextPendingKeys = new HashSet<>(pendingCallModes);
+            Set<CallMode> triggeredCalls = new HashSet<>();
+            iterate(orderingChoice.cyclicConcludableBounds).forEachRemaining(concludableBounds -> {
+                triggeredCalls.addAll(triggeredCalls(concludableBounds.first(), concludableBounds.second(), Optional.of(conjunctionNode.cyclicDependencies(concludableBounds.first()))));
+            });
+            iterate(triggeredCalls).filter(call -> !choices.containsKey(call)).forEachRemaining(nextPendingKeys::add);
+            ComponentPlan newPlan = componentPlanSearch(root, nextPendingKeys, choices);
+
+            double newPlanCost = newPlan.cost(root);
+            if (bestPlan == null || newPlanCost < bestPlanCost) {
+                bestPlan = newPlan;
+                bestPlanCost = newPlanCost;
+            }
+
+            choices.remove(mode);
+        }
+        assert bestPlan != null;
+        return bestPlan;
+    }
+
+    private void initialiseOrderingDependencies(ConjunctionNode conjunctionNode, List<Resolvable<?>> ordering, Set<Variable> inputBounds) {
+        Set<Variable> currentBounds = new HashSet<>(inputBounds);
+        for (Resolvable<?> resolvable : ordering) {
+            Set<Variable> resolvableBounds = iterate(consideredVariables(resolvable)).filter(currentBounds::contains).toSet();
+            initialiseDependencies(conjunctionNode, resolvable, resolvableBounds);
+            if (!resolvable.isNegated()) currentBounds.addAll(consideredVariables(resolvable));
```

Comment:
if you're going to do an operation twice, `consideredVariables(resolvable)` it's normally better to save it to a variable to avoid double-computation